### PR TITLE
[JSC] Implement Symbols as WeakMap keys

### PR DESCRIPTION
**Pull request by Constellation**
#### 290c0b91225ac430d66b4fff9a4f6606b063ba5d
<pre>
[JSC] Implement Symbols as WeakMap keys
<a href="https://bugs.webkit.org/show_bug.cgi?id=243483">https://bugs.webkit.org/show_bug.cgi?id=243483</a>

Reviewed by Ross Kirsling and Saam Barati.

This patch implements stage-3 proposal &quot;Symbols as WeakMap keys&quot;[1].
Previously, WeakMap, WeakSet, FinalizationRegistry, and WeakRef are only accepting objects as keys.
But this proposal extends it to accept non-registered Symbols too.

The runtime implementation is just changed from using JSObject* to JSCell*. And we check whether the
given value is appropriate by using canBeHeldWeakly function (specified in the spec) instead of `isObject()`.

A bit complicated part is DFG / FTL handling for WeakMap / WeakSet functions. We extend it to accept SymbolUse
and CellUse. And we emit appropriate code for that, and modifying clobberizing rules and doesGC rules too for
these new UseKind.

[1]: <a href="https://github.com/tc39/proposal-symbols-as-weakmap-keys">https://github.com/tc39/proposal-symbols-as-weakmap-keys</a>

* JSTests/stress/finalization-registry-registered-symbol.js: Added.
(shouldThrow):
* JSTests/stress/v8-cleanup-from-different-realm-symbol.js: Added.
(let.timeout_func):
* JSTests/stress/v8-cleanup-proxy-from-different-realm-symbol.js: Added.
(let.timeout_func):
* JSTests/stress/v8-finalization-registry-basics-symbol.js: Copied from JSTests/stress/v8-finalization-registry-basics.js.
(TestConstructFinalizationRegistry):
(TestFinalizationRegistryConstructorCallAsFunction):
(TestConstructFinalizationRegistryCleanupNotCallable):
(TestConstructFinalizationRegistryWithNonCallableProxyAsCleanup):
(TestRegisterTargetAndHoldingsSameValue):
(TestRegisterWithoutFinalizationRegistry):
(TestUnregisterWithNonExistentKey):
(TestUnregisterWithNonFinalizationRegistry):
(TestWeakRefConstructorWithNonObject):
* JSTests/stress/v8-finalization-registry-basics.js:
(TestWeakRefConstructorWithNonObject):
* JSTests/stress/v8-finalizationregistry-and-weakref-symbol.js: Added.
(let.cleanup):
(setTimeout):
* JSTests/stress/v8-finalizationregistry-keeps-holdings-alive-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-finalizationregistry-scheduled-for-cleanup-multiple-times-symbol.js: Added.
(let.cleanup0):
(let.cleanup1):
(let.timeout_func):
* JSTests/stress/v8-multiple-dirty-finalization-registries-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-reentrant-gc-from-cleanup-symbol.js: Added.
(let.reentrant_gc):
(setTimeout):
* JSTests/stress/v8-stress-finalizationregistry-dirty-enqueue-symbol.js: Added.
(i.registries.push.new.FinalizationRegistry):
(registries.forEach):
* JSTests/stress/v8-undefined-holdings-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-after-cleanup-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-before-cleanup-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-called-twice-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-inside-cleanup2-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-inside-cleanup3-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-many-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-weak-unregistertoken-symbol.js: Added.
(FR.new.FinalizationRegistry):
(tryAgain):
* JSTests/stress/weak-map-constructor.js:
* JSTests/stress/weak-map-registered-symbol.js: Added.
(shouldBe):
(shouldThrow):
* JSTests/stress/weak-map-symbol.js: Added.
(shouldBe):
(test):
* JSTests/stress/weak-map-various-one-site.js: Added.
(shouldThrow):
(getTest):
(hasTest):
(setTest):
(deleteTest):
(i.test.set shouldBe):
(i.test):
* JSTests/stress/weak-map-various.js: Added.
(shouldBe):
* JSTests/stress/weak-ref-registered-symbol.js: Added.
(shouldThrow):
* JSTests/stress/weak-set-constructor.js:
* JSTests/stress/weak-set-registered-symbol.js: Added.
(shouldBe):
(shouldThrow):
(test.set add):
* JSTests/stress/weak-set-symbol.js: Added.
(shouldBe):
* JSTests/stress/weak-set-various-one-site.js: Added.
(shouldThrow):
(set shouldThrow):
* JSTests/stress/weak-set-various.js: Added.
(shouldBe):
* JSTests/test262/config.yaml:
* LayoutTests/js/dom/basic-weakmap-expected.txt:
* LayoutTests/js/dom/basic-weakset-expected.txt:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::registerTarget):
(JSC::JSFinalizationRegistry::unregister):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSWeakMap.h:
* Source/JavaScriptCore/runtime/JSWeakMapInlines.h:
(JSC::JSWeakMap::set):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.cpp:
(JSC::JSWeakObjectRef::finishCreation):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.h:
* Source/JavaScriptCore/runtime/WeakMapConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakMapImpl.cpp:
(JSC::WeakMapImpl&lt;WeakMapBucket&gt;::takeSnapshotInternal):
(JSC::WeakMapImpl&lt;WeakMapBucket&lt;WeakMapBucketDataKey&gt;&gt;::takeSnapshot):
(JSC::WeakMapImpl&lt;WeakMapBucket&lt;WeakMapBucketDataKeyValue&gt;&gt;::takeSnapshot):
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapBucket::setKey):
(JSC::WeakMapBucket::key const):
(JSC::WeakMapBucket::deletedKey):
(JSC::WeakMapImpl::get):
(JSC::WeakMapImpl::findBucket):
(JSC::WeakMapImpl::has):
(JSC::WeakMapImpl::remove):
(JSC::WeakMapImpl::canUseBucket):
(JSC::WeakMapImpl::addInternal):
(JSC::WeakMapImpl::findBucketAlreadyHashed):
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::jsWeakMapHash):
(JSC::canBeHeldWeakly):
(JSC::WeakMapImpl&lt;WeakMapBucket&gt;::add):
* Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
(JSC::WeakMapPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakMapPrototype.h:
* Source/JavaScriptCore/runtime/WeakObjectRefConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetPrototype.cpp:
(JSC::WeakSetPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetPrototype.h:

Canonical link: <a href="https://commits.webkit.org/253135@main">https://commits.webkit.org/253135@main</a>
</pre>

### CODE REVIEW

File: JSTests/stress/v8-cleanup-from-different-realm-symbol.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2018 the V8 project authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Flags: --harmony-weak-refs --expose-gc --noincremental-marking
+
+load("./resources/v8-mjsunit.js", "caller relative");
```
Comment (rkirsling) [inline JSTests/stress/v8-cleanup-from-different-realm-symbol.js:7]: I guess it would have to be done in a separate patch, but what do you think about moving `stress/v8-*.js` to a `v8/` directory?

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp [hunk 2]
```diff
@@ -31,7 +31,7 @@
 
 namespace JSC {
 
-const ASCIILiteral WeakMapNonObjectKeyError { "Attempted to set a non-object key in a WeakMap"_s };
+const ASCIILiteral WeakMapInvalidKeyError { "Attempted to set neither an object nor a non-registered symbol key in a WeakMap"_s };
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: This phrasing is pretty hard to read, let's make it more like:
```suggestion
const ASCIILiteral WeakMapInvalidKeyError { "WeakMap keys must be objects or non-registered symbols"_s };
```

File: Source/JavaScriptCore/runtime/WeakSetPrototype.cpp [hunk 3]
```diff
@@ -33,7 +33,7 @@
 
 namespace JSC {
 
-const ASCIILiteral WeakSetNonObjectValueError { "Attempted to add a non-object value to a WeakSet"_s };
+const ASCIILiteral WeakSetInvalidValueError { "Attempted to add neither an object nor a non-registered symbol value to a WeakSet"_s };
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/WeakSetPrototype.cpp]: Ditto.
```suggestion
const ASCIILiteral WeakSetInvalidValueError { "WeakSet values must be objects or non-registered symbols."_s };
```

File: JSTests/test262/config.yaml [hunk 4]
```diff
@@ -372,3 +372,12 @@ skip:
     - test/built-ins/Temporal/Instant/prototype/until/instant-string.js
     - test/built-ins/Temporal/PlainTime/prototype/since/balance-negative-time-units.js
     - test/built-ins/Temporal/PlainTime/prototype/until/balance-negative-time-units.js
+
+    # Symbols as WeakMap keys proposal breaks existing tests https://github.com/tc39/proposal-symbols-as-weakmap-keys
```
Comment (rkirsling) [inline JSTests/test262/config.yaml:376]: Should we report this in a GH issue somewhere?

File: LayoutTests/js/dom/basic-weakmap-expected.txt [hunk 5]
```diff
@@ -7,13 +7,13 @@ PASS WeakMap instanceof WeakMap is false
...
 PASS new WeakMap() instanceof WeakMap is true
 PASS WeakMap() threw exception TypeError: calling WeakMap constructor without new is invalid.
-PASS map.set(0, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(0.5, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set('foo', 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(true, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(false, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(null, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(undefined, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
+PASS map.set(0, 1) threw exception TypeError: Attempted to set neither an object nor a non-registered symbol key in a WeakMap.
```
Comment (rkirsling) [inline LayoutTests/js/dom/basic-weakmap-expected.txt]: How about moving these LayoutTests to JSTests/stress? :smile:

File: Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp [hunk 6]
```diff
@@ -97,12 +98,12 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncFinalizationRegistryUnregister, (JSGlobalObjec
     auto* group = getFinalizationRegistry(vm, globalObject, callFrame->thisValue());
     RETURN_IF_EXCEPTION(scope, { });
 
-    if (auto* token = jsDynamicCast<JSObject*>(callFrame->argument(0))) {
-        bool result = group->unregister(vm, token);
-        return JSValue::encode(jsBoolean(result));
-    }
+    JSValue token = callFrame->argument(0);
+    if (UNLIKELY(!canBeHeldWeakly(token)))
+        return throwVMTypeError(globalObject, scope, "unregister requires an object or a non-registered symbol is the unregistration token"_s);
```
Comment (rkirsling) [inline Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp]:
```suggestion
        return throwVMTypeError(globalObject, scope, "unregister requires an object or a non-registered symbol as the unregistration token"_s);
```

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 7]
```diff
@@ -15290,7 +15315,13 @@ void SpeculativeJIT::compileWeakMapGet(Node* node)
 
     SpeculateCellOperand key(this, node->child2());
     GPRReg keyGPR = key.gpr();
-    speculateObject(node->child2(), keyGPR);
+
+    // We are not checking whether Symbol is registered one or not, but it is OK since Symbol and SymbolImpl are one-on-one.
```
Comment (rkirsling) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: nit
```suggestion
    // We are not checking whether Symbol is registered one or not, but it is OK since Symbol and SymbolImpl are one-to-one.
```

**Review by rkirsling — APPROVED** [review]
r=me, though I'm not sure I fully understand the clobbering situation.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp [hunk 8]
```diff
@@ -2564,6 +2564,11 @@ class FixupPhase : public Phase {
                 break;
             }
 
+            if (node->child1()->shouldSpeculateSymbol()) {
+                fixEdge<SymbolUse>(node->child1());
+                break;
+            }
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGFixupPhase.cpp]: This code already exists above :-)

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp [hunk 9]
```diff
@@ -2600,7 +2605,10 @@ class FixupPhase : public Phase {
                 fixEdge<WeakSetObjectUse>(node->child1());
             else
                 RELEASE_ASSERT_NOT_REACHED();
-            fixEdge<ObjectUse>(node->child2());
+            if (node->child2()->shouldSpeculateObject())
+                fixEdge<ObjectUse>(node->child2());
+            else if (node->child2()->shouldSpeculateSymbol())
+                fixEdge<SymbolUse>(node->child2());
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:2606]: since this is relying on BytecodeParser laying down CellUse, I suggest we add a validation rule for WeakMapGet where we assert we have the right kinds of speculations on edges.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp [hunk 10]
```diff
@@ -2619,14 +2627,16 @@ class FixupPhase : public Phase {
 
         case WeakSetAdd: {
             fixEdge<WeakSetObjectUse>(node->child1());
-            fixEdge<ObjectUse>(node->child2());
+            if (node->child2()->shouldSpeculateObject())
+                fixEdge<ObjectUse>(node->child2());
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:2626]: same comment about validation rule

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp [hunk 11]
```diff
@@ -2619,14 +2627,16 @@ class FixupPhase : public Phase {
...
             fixEdge<Int32Use>(node->child3());
             break;
         }
 
         case WeakMapSet: {
             fixEdge<WeakMapObjectUse>(m_graph.varArgChild(node, 0));
-            fixEdge<ObjectUse>(m_graph.varArgChild(node, 1));
+            if (m_graph.varArgChild(node, 1)->shouldSpeculateObject())
+                fixEdge<ObjectUse>(m_graph.varArgChild(node, 1));
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:2634]: Same comment about validation rule

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp [hunk 12]
```diff
@@ -76,7 +76,9 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncWeakMapDelete, (JSGlobalObject* globalObject,
     if (!map)
         return JSValue::encode(jsUndefined());
     JSValue key = callFrame->argument(0);
-    return JSValue::encode(jsBoolean(key.isObject() && map->remove(asObject(key))));
+    if (UNLIKELY(!canBeHeldWeakly(key)))
+        return JSValue::encode(jsBoolean(false));
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: Is it worth doing the extra uid load and isDerred load? Maybe it's faster to just do the hash table lookup since it's unlikely that anyone is a registered symbol. And we're guaranteed to not have the key in the set if it's registered.

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp [hunk 13]
```diff
@@ -85,9 +87,9 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncWeakMapGet, (JSGlobalObject* globalObject, Cal
     if (!map)
         return JSValue::encode(jsUndefined());
     JSValue key = callFrame->argument(0);
-    if (!key.isObject())
+    if (UNLIKELY(!canBeHeldWeakly(key)))
         return JSValue::encode(jsUndefined());
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: ditto as above, is it even worth the isRegistered check here since it'll usually be false and just cost us extra loads

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp [hunk 14]
```diff
@@ -96,7 +98,9 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncWeakMapHas, (JSGlobalObject* globalObject, Cal
     if (!map)
         return JSValue::encode(jsUndefined());
     JSValue key = callFrame->argument(0);
-    return JSValue::encode(jsBoolean(key.isObject() && map->has(asObject(key))));
+    if (UNLIKELY(!canBeHeldWeakly(key)))
+        return JSValue::encode(jsBoolean(false));
```
Comment (saambarati) [inline Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: ditto. Probably this happens in other access functions too, but I won't annotate them all

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp (hunk 12 above)
Comment (saambarati) [reply Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: Maybe we want some function for this that's different than canBeHeldWeakly since we use that necessarily along the add path

File: Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp [hunk 15]
```diff
@@ -12639,65 +12639,99 @@ IGNORE_CLANG_WARNINGS_END
...
+            break;
+        }
+        case CellUse: {
+            LValue key = lowCell(m_node->child2());
+            if (m_node->child1().useKind() == WeakMapObjectUse)
+                setJSValue(vmCall(Int64, operationWeakMapGet, m_vmValue, weakMap, key, hash));
+            else
+                setJSValue(vmCall(Int64, operationWeakSetGet, m_vmValue, weakMap, key, hash));
+            break;
+        }
```
Comment (saambarati) [inline Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp]: Why slow path here? Why can't we run the above path?

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp [hunk 16]
```diff
@@ -15256,6 +15256,31 @@ void SpeculativeJIT::compileMapSet(Node* node)
...
+        JSValueRegsFlushedCallResult result(this);
+        JSValueRegs resultRegs = result.regs();
+        if (node->child1().useKind() == WeakMapObjectUse)
+            callOperation(operationWeakMapGet, resultRegs, TrustedImmPtr(&vm()), weakMapGPR, keyGPR, hashGPR);
+        else
+            callOperation(operationWeakSetGet, resultRegs, TrustedImmPtr(&vm()), weakMapGPR, keyGPR, hashGPR);
+        m_jit.exceptionCheck();
+        jsValueResult(resultRegs, node);
+        return;
+    }
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: ditto as comment about FTL code, why can't we run the fast path here?
If it's some bad cell value, like a string or a register symbol, we won't find it in the hash map 
Maybe just a comment to explain why it's ok

**Review by saambarati — APPROVED** [review]

File: Source/JavaScriptCore/dfg/DFGClobberize.h [hunk 17]
```diff
@@ -1952,7 +1952,8 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
         Edge& keyEdge = node->child2();
         AbstractHeapKind heap = (mapEdge.useKind() == WeakMapObjectUse) ? JSWeakMapFields : JSWeakSetFields;
         read(heap);
-        def(HeapLocation(WeakMapGetLoc, heap, mapEdge, keyEdge), LazyNode(node));
+        if (keyEdge.useKind() == ObjectUse || keyEdge.useKind() == SymbolUse)
+            def(HeapLocation(WeakMapGetLoc, heap, mapEdge, keyEdge), LazyNode(node));
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGClobberize.h]: Why can't we always def here, even for CellUse? If we have a string or registered symbol, its result is false

File: Source/JavaScriptCore/dfg/DFGClobberize.h [hunk 18]
```diff
@@ -1975,17 +1976,23 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
     case WeakSetAdd: {
         Edge& mapEdge = node->child1();
         Edge& keyEdge = node->child2();
-        write(JSWeakSetFields);
-        def(HeapLocation(WeakMapGetLoc, JSWeakSetFields, mapEdge, keyEdge), LazyNode(keyEdge.node()));
+        if (keyEdge.useKind() == ObjectUse) {
+            write(JSWeakSetFields);
+            def(HeapLocation(WeakMapGetLoc, JSWeakSetFields, mapEdge, keyEdge), LazyNode(keyEdge.node()));
+        } else
```
Comment (saambarati) [inline Source/JavaScriptCore/dfg/DFGClobberize.h]: This one makes sense to me, since we might throw if it's a registered symbol. But I wonder if we can still CSE it? Also, maybe we can add a speculation like NonRegisterdSymbol at some point.

File: JSTests/stress/v8-cleanup-from-different-realm-symbol.js (hunk 1 above)
Comment (Constellation) [reply JSTests/stress/v8-cleanup-from-different-realm-symbol.js:7]: Yeah, that would be good change in a subsequent patch.

File: JSTests/test262/config.yaml (hunk 4 above)
Comment (Constellation) [reply JSTests/test262/config.yaml:376]: Maybe, we should convert it to actual testings. But for now, let's just skip them.

File: LayoutTests/js/dom/basic-weakmap-expected.txt (hunk 5 above)
Comment (Constellation) [reply LayoutTests/js/dom/basic-weakmap-expected.txt]: I would like to defer these tasks since we should do that for all tests in `LayoutTests/js` at the same time (and if it is tied to DOM, then we should move them to `LayoutTests/js/dom`) :)

File: Source/JavaScriptCore/dfg/DFGClobberize.h (hunk 17 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGClobberize.h]: Yeah, let's def it.

File: Source/JavaScriptCore/dfg/DFGClobberize.h (hunk 18 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGClobberize.h]: I think probably we can just `read(World)` and `write(SideState)` too.
Then we will not reorder this node with the other one which writes something. (ensuring that exception is thrown in user observable way).
So I think we can still define heap location. Changed.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp (hunk 8 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGFixupPhase.cpp]: Oops, removed.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp (hunk 9 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:2606]: Sounds good.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp (hunk 10 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:2626]: Sounds good.

File: Source/JavaScriptCore/dfg/DFGFixupPhase.cpp (hunk 11 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:2634]: Ditto.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp (hunk 16 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: Yeah, right. Removed this.

File: Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp (hunk 7 above)
Comment (Constellation) [reply Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp]: Fixed.

File: Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp (hunk 15 above)
Comment (Constellation) [reply Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp]: Yeah, removed.

File: Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp (hunk 6 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp]: Fixed.

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp (hunk 2 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: Looks good. Changed.

File: Source/JavaScriptCore/runtime/WeakSetPrototype.cpp (hunk 3 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/WeakSetPrototype.cpp]: Changed.

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp (hunk 12 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: Probably, we should just check it is a cell. That's enough.

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp (hunk 13 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: Yeah, changed it to `isCell()` check.

File: Source/JavaScriptCore/runtime/WeakMapPrototype.cpp (hunk 14 above)
Comment (Constellation) [reply Source/JavaScriptCore/runtime/WeakMapPrototype.cpp]: Changed.
